src/index.tsx

```diff
@@ -0,0 +1,122 @@
+import { useEffect } from 'react';
+import {
+	PingType,
+	IsPollingType,
+	GetPollingConfigType,
+	UseOnlineEffectType,
+} from './types';
+
+export const needsPolling: IsPollingType = navigator => {
+	// these browsers don't fully support navigator.onLine, so we need to use a polling backup
+	const unsupportedUserAgentsPattern = /Windows.*Chrome|Windows.*Firefox|Linux.*Chrome/;
+
+	if (
+		typeof navigator !== 'undefined' &&
+		unsupportedUserAgentsPattern.test(navigator.userAgent)
+	) {
+		return true;
+	}
+
+	return false;
+};
+
+export const ping: PingType = ({ url, timeout }) => {
+	return new Promise(resolve => {
+		const isOnline = () => resolve(true);
+		const isOffline = () => resolve(false);
+
+		const xhr = new XMLHttpRequest();
+
+		xhr.onerror = isOffline;
+		xhr.ontimeout = isOffline;
+		xhr.onreadystatechange = () => {
+			if (xhr.readyState === xhr.HEADERS_RECEIVED) {
+				if (xhr.status) {
+					isOnline();
+				} else {
+					isOffline();
+				}
+			}
+		};
+
+		xhr.open('HEAD', url);
+		xhr.timeout = timeout;
+		xhr.send();
+	});
+};
+
+const getPollingConfigs: GetPollingConfigType = (
+	pollingConfig,
+	needsPolling
+) => {
+	const defaultConfig = {
+		enabled: true,
+		url: 'https://ipv4.icanhazip.com/',
+		timeout: 5000,
+		interval: 5000,
+	};
+
+	if (
+		(typeof pollingConfig === 'object' && pollingConfig.enabled === true) ||
+		(needsPolling === true && typeof pollingConfig === 'object')
+	) {
+		return { ...defaultConfig, ...pollingConfig };
+	} else if (pollingConfig === true || needsPolling) {
+		return { ...defaultConfig };
+	} else {
+		return { enabled: false };
+	}
+};
+
+export const useOnlineEffect: UseOnlineEffectType = (
+	callback,
```

Comment:
Better than using a callback would be use useState.
```typescript
const [isOnline, setIsOnline] = useState(true);

const goOnline = () => {
    setIsOnline(true);
};

const goOffline = () => {
    setIsOnline(false);
};

...
setInterval(() => {
        ping(pingConfig).then(setIsOnline);
      }, interval);
...

return isOnline;
```
A user can then just use it like:
```typescript
const isOnline = useOnlineEffect();

useEffect(() => {
  if (isOnline) { 
    console.debug("we're online");
  } else {
    console.warn("we're offline");
  }
}, [isOnline])

Reply:
If callback is still wanted, you could add
```typescript
if (callback) {
  useEffect(() => {
    callback(isOnline);
  }, [isOnline])
}
````
to have both options.